policybot/pkg/gh/ghs.go

```diff
@@ -129,3 +131,20 @@ func (ghs *GitHubState) ReadPullRequestReview(org string, repo string, issue str
 
 	return ghs.store.ReadPullRequestReviewByID(org, repo, issue, prReview)
 }
+
+// ReadIssueBySQL returns issue based on the SQL query.
```

Comment:
I think this is a layering violation. The interface storage/store.go hides the details of the underlying storage layer such as the specific database engine, the specific table schemas, etc. Exposing this method means that storage is no longer abstract.

I think we need to have dedicated fixed-function query functions exposed from storage/store.go which encapsulates the specific details of the query and return data in terms of the storage/types.go objects.

Reply:
I don't quite agree... I figured more bots need more flexible SQL query later on and hardly think we'll support more than Spanner..

but I can change if you think this is really needed. If we really change, a couple of questions,

- Who holds the SQL query definition? I think per bot package makes more sense (flakechaser here).
- Which package owns the Spanner row post-processing?

Reply:
First, let me say that I'm not an expert in how to interact with databases, this is my first larger scale interaction with DBs. So take my opinion in that context.

I consulted a few best practices docs and looked at some existing code I found out there and the general model seems to be to try to encapsulate all storage interactions behind an API via a DAO.  This clearly involves a tradeoff:

- Either the DAO needs to understand enough of the business logic to expose all the necessary queries.

- Or the inner design of the database is exposed to the upper layers of the stack so they can directly implement custom queries.

- Or something in the middle, where the DAO exposes some sort of abstract query model that's independent of the underlying storage model.

So I opted to go with the common wisdom of encapsulating all storage access into the DAO. So can we start with that approach and if that doesn't work out we can always break the abstraction layer.

The implication here is to add some dedicated QueryXXX methods to the storage.Store interface that fully contain the interaction with the database. The model I started to use here is to provide a callback function to the Query method which gets invoked for every row that the query produces.

Now that I've finally got all the syncing stuff out of the way, I'm going to start writing queries this weekend to capture the issue, pr, maintainer, and member SLOs. I'm hoping to adopt this pattern.

Thanks.

Reply:
Thanks for the detailed explanation, this is much clearer. Yeah, I do recall that in my previous experience, where different components use database/bigtable schema as implicit interface for information sharing. The end result makes it impossible to change the database schema.

The bots here is not that complicated, but I guess it's fine to to adopt simple approach. I'll change by adding a `ReadTestFlakesIssuse(days int)` and implemented that in spanner layer.

Reply:
Changed.